### ISSUE TITLE
Crash on failed purchase in async code on iOS 14

### ISSUE THREAD

**Issue by anivaros** [RevenueCat/purchases-ios#1868]
- [x] I have updated Purchases SDK to the latest version
- [x] I have read the [Contribution Guidelines](https://github.com/RevenueCat/purchases-ios/blob/main/Development/CONTRIBUTING.md)
- [x] I have searched the [Community](https://community.revenuecat.com)
- [x] I have read [docs.revenuecat.com](https://docs.revenuecat.com/)
- [x] I have searched for [existing Github issues](https://github.com/RevenueCat/purchases-ios/issues)

**Describe the bug**
Lib crashes on iOS 14 after error in purchase (like cancel)

1. Environment
   1. Platform: iOS
   2. SDK version: 4.10.3
   3. StoreKit 2 (_enabled with `useStoreKit2IfEnabled`_) (Y/N): **Y** (but on iOS 14 of course **N**)
   4. OS version: iOS 14
   5. Xcode version: 13.4.1
   6. 100% reproducible 
2. 
> 2022-08-27 11:05:05.299083+0300 CookingRecipes[99925:2979178] [Purchases] - ERROR: 🍎‼️ Purchase was cancelled.
RevenueCat/Purchases+async.swift:69: Fatal error: Unexpected nil value and nil error
2022-08-27 11:05:05.299627+0300 CookingRecipes[99925:2979178] RevenueCat/Purchases+async.swift:69: Fatal error: Unexpected nil value and nil error

3. Make purchase, cancel on iOS purchase screen
4. 
> 0  libswiftCore.dylib             0x37acc _assertionFailure(_:_:file:line:flags:) + 1532
1  CookingRecipes                 0x1ff5bc specialized Result.init(_:_:file:line:) + 24 (Result+Extensions.swift:24)
2  CookingRecipes                 0x2437bc closure #1 in closure #1 in Purchases.purchaseAsync(product:) + 4341921724
3  CookingRecipes                 0x244c7c partial apply for closure #1 in closure #1 in Purchases.purchaseAsync(package:promotionalOffer:) + 4341927036
4  CookingRecipes                 0x295498 thunk for @escaping @callee_guaranteed (@guaranteed StoreTransaction?, @guaranteed CustomerInfo?, @guaranteed Error?, @unowned Bool) -> () + 4342256792 (<compiler-generated>:4342256792)
5  CookingRecipes                 0x2accf8 partial apply for thunk for @escaping @callee_unowned @convention(block) (@unowned StoreTransaction?, @unowned CustomerInfo?, @unowned NSError?, @unowned ObjCBool) -> () + 4342353144 (<compiler-generated>:4342353144)
6  CookingRecipes                 0x2a6484 closure #2 in PurchasesOrchestrator.purchase(sk1Product:payment:package:completion:) + 352 (PurchasesOrchestrator.swift:352)
7  CookingRecipes                 0x2acd34 partial apply for thunk for @escaping @callee_guaranteed (@guaranteed StoreTransaction?, @guaranteed CustomerInfo?, @guaranteed Error?, @unowned Bool) -> () + 4342353204 (<compiler-generated>:4342353204)
8  CookingRecipes                 0x2add28 partial apply for thunk for @escaping @callee_guaranteed (@in_guaranteed StoreTransaction?, @in_guaranteed CustomerInfo?, @in_guaranteed Error?, @in_guaranteed Bool) -> (@out ()) + 4342357288 (<compiler-generated>:4342357288)
9  CookingRecipes                 0x2a83c0 PurchasesOrchestrator.handleFailedTransaction(_:) + 4342334400 (<compiler-generated>:4342334400)
10 CookingRecipes                 0x2a80c0 PurchasesOrchestrator.storeKitWrapper(_:updatedTransaction:) + 514 (PurchasesOrchestrator.swift:514)
11 CookingRecipes                 0x2b9260 specialized StoreKitWrapper.paymentQueue(_:updatedTransactions:) + 137 (StoreKitWrapper.swift:137)
12 CookingRecipes                 0x2b8de4 @objc StoreKitWrapper.paymentQueue(_:updatedTransactions:) + 4342402532
13 StoreKit                       0x10e6c __63-[SKPaymentQueue _processTransactionDict:forTransaction:error:]_block_invoke_2 + 116

**Comment by RCGitBot** [comment]
👀 [SDKONCALL-110] We've just linked this issue to our internal tracker and notified the team. Thank you for reporting, we're checking this out!

**Comment by FelixLisczyk** [comment]
I can reproduce this issue locally with RevenueCat SDK 4.10.2 and 4.10.3 on iOS 15+ as long as StoreKit 2 is **not** enabled. RevenueCat SDK 4.10.1 works fine.

However I'm also getting iOS 15 production crashes with RevenueCat SDK 4.10.2 where StoreKit 2 **is enabled**. Please see the crash log from App Center below. I will downgrade to 4.10.1 for the time being.

```
MAIN THREAD - CRASHED
libswiftCore.dylib
Swift._assertionFailure(_: Swift.StaticString, _: Swift.String, file: Swift.StaticString, line: Swift.UInt, flags: Swift.UInt32) -> Swift.Never
RevenueCat_76EF69325FE509FC_PackageProduct
function signature specialization <Arg[3] = Dead, Arg[4] = Owned To Guaranteed> of function signature specialization <Arg[1] = [Closure Propagated : $s10RevenueCat9PurchasesC13purchaseAsync7packageAA16StoreTransactionCSg11transaction_AA12CustomerInfoC08customerK0Sb13userCancelledtAA7PackageC_tYaKFyScCyAhI_AkLSbAMts5Error_pGXEfU_yAH_AKSgsAP_pSgSbtcfU_ASyXEfu_, Argument Types : [Swift.Error?Swift.Bool]> of generic specialization <RevenueCat.CustomerInfo, Swift.Error> of (extension in RevenueCat):Swift.Result.init(_: A?, _: @autoclosure () -> B?, file: Swift.StaticString, line: Swift.UInt) -> Swift.Result<A, B> Result+Extensions.swift:24
RevenueCat_76EF69325FE509FC_PackageProduct
$s10RevenueCat9PurchasesC13purchaseAsync7productAA16StoreTransactionCSg11transaction_AA12CustomerInfoC08customerK0Sb13userCancelledtAA0G7ProductC_tYaKFyScCyAhI_AkLSbAMts5Error_pGXEfU_yAH_AKSgsAP_pSgSbtcfU_Tm
RevenueCat_76EF69325FE509FC_PackageProduct
$s10RevenueCat9PurchasesC13purchaseAsync7package16promotionalOfferAA16StoreTransactionCSg11transaction_AA12CustomerInfoC08customerM0Sb13userCancelledtAA7PackageC_AA011PromotionalH0CtYaKFyScCyAiJ_AlMSbANts5Error_pGXEfU_yAI_ALSgsAS_pSgSbtcfU_TATm
RevenueCat_76EF69325FE509FC_PackageProduct
reabstraction thunk helper from @escaping @callee_guaranteed (@guaranteed RevenueCat.StoreTransaction?, @guaranteed RevenueCat.CustomerInfo?, @guaranteed Swift.Error?, @unowned Swift.Bool) -> () to @escaping @callee_unowned @convention(block) (@unowned RevenueCat.StoreTransaction?, @unowned RevenueCat.CustomerInfo?, @unowned __C.NSError?, @unowned ObjectiveC.ObjCBool) -> () <compiler-generated>:0
RevenueCat_76EF69325FE509FC_PackageProduct
partial apply forwarder for reabstraction thunk helper from @escaping @callee_unowned @convention(block) (@unowned RevenueCat.StoreTransaction?, @unowned RevenueCat.CustomerInfo?, @unowned __C.NSError?, @unowned ObjectiveC.ObjCBool) -> () to @escaping @callee_guaranteed (@guaranteed RevenueCat.StoreTransaction?, @guaranteed RevenueCat.CustomerInfo?, @guaranteed Swift.Error?, @unowned Swift.Bool) -> () <compiler-generated>:0
RevenueCat_76EF69325FE509FC_PackageProduct
closure #2 (RevenueCat.StoreTransaction?, RevenueCat.CustomerInfo?, Swift.Error?, Swift.Bool) -> () in RevenueCat.PurchasesOrchestrator.purchase(sk1Product: __C.SKProduct, payment: __C.SKMutablePayment, package: RevenueCat.Package?, completion: (RevenueCat.StoreTransaction?, RevenueCat.CustomerInfo?, Swift.Error?, Swift.Bool) -> ()) -> () PurchasesOrchestrator.swift:351
RevenueCat_76EF69325FE509FC_PackageProduct
partial apply forwarder for reabstraction thunk helper from @escaping @callee_guaranteed (@guaranteed RevenueCat.StoreTransaction?, @guaranteed RevenueCat.CustomerInfo?, @guaranteed Swift.Error?, @unowned Swift.Bool) -> () to @escaping @callee_guaranteed (@in_guaranteed RevenueCat.StoreTransaction?, @in_guaranteed RevenueCat.CustomerInfo?, @in_guaranteed Swift.Error?, @in_guaranteed Swift.Bool) -> (@out ()) <compiler-generated>:0
RevenueCat_76EF69325FE509FC_PackageProduct
partial apply forwarder for reabstraction thunk helper from @escaping @callee_guaranteed (@in_guaranteed RevenueCat.StoreTransaction?, @in_guaranteed RevenueCat.CustomerInfo?, @in_guaranteed Swift.Error?, @in_guaranteed Swift.Bool) -> (@out ()) to @escaping @callee_guaranteed (@guaranteed RevenueCat.StoreTransaction?, @guaranteed RevenueCat.CustomerInfo?, @guaranteed Swift.Error?, @unowned Swift.Bool) -> () <compiler-generated>:0
RevenueCat_76EF69325FE509FC_PackageProduct
RevenueCat.PurchasesOrchestrator.(handleFailedTransaction in _B8F471F86775DD0D24EF3C9BB319BEBB)(__C.SKPaymentTransaction) -> () PurchasesOrchestrator.swift:593
RevenueCat_76EF69325FE509FC_PackageProduct
RevenueCat.PurchasesOrchestrator.storeKitWrapper(_: RevenueCat.StoreKitWrapper, updatedTransaction: __C.SKPaymentTransaction) -> () PurchasesOrchestrator.swift:514
RevenueCat_76EF69325FE509FC_PackageProduct
function signature specialization <Arg[0] = Dead> of RevenueCat.StoreKitWrapper.paymentQueue(_: __C.SKPaymentQueue, updatedTransactions: [__C.SKPaymentTransaction]) -> () StoreKitWrapper.swift:0
RevenueCat_76EF69325FE509FC_PackageProduct
merged @objc RevenueCat.StoreKitWrapper.paymentQueue(_: __C.SKPaymentQueue, updatedTransactions: [__C.SKPaymentTransaction]) -> ()
StoreKit
-[SKPaymentQueue _processTransactionDict:forTransaction:error:]
```

**Comment by NachoSoto** [comment]
This is definitely because of #1841.

Does this happen on device or simulator?
There must be a code path where we only send the cancellation error and not the `CustomerInfo`.

It's so unfortunate that there's no way to write tests for cancellations with StoreKit.

**Comment by anivaros** [comment]
> Does this happen on device or simulator?

This happens everywhere, on simulator and on production 😔

**Comment by NachoSoto** [comment]
I won't be able to look at this until tomorrow. In the mean time it would be safe to revert that commit.

**Comment by NachoSoto** [comment]
Ok thanks for the detailed report!
I will look at this tomorrow. In the meantime I recommend downgrading to the earlier version.

**Comment by NachoSoto** [comment]
Okay confirmed this is an SK2-only issue:
- https://github.com/RevenueCat/purchases-ios/blob/4.10.3/Sources/Purchasing/Purchases/PurchasesOrchestrator.swift#L594-L596
- https://github.com/RevenueCat/purchases-ios/blob/4.10.3/Sources/Purchasing/Purchases/PurchasesOrchestrator.swift#L614-L616

We could make the logic in #1841 smarter to avoid the crash, but the real fix should be in `PurchasesOrchestrator` so that we have consistent return values and we always have a `CustomerInfo` (or not) in case of cancellations.
Right now depending on how it's cancelled / failed / SK1 / SK2 we're returning different combinations of values.

**Comment by NachoSoto** [comment]
Here's a summary of the inconsistency. TLDR, the crash is happening only when using the `async` APIs and either SK2 is disabled, or the app is running on an SK1-only device.

### Before #1841:
- SK1 / completion-block: no `CustomerInfo` + cancelled + `ErrorCode.purchaseCancelledError`
- SK1 / `async`: thrown `ErrorCode.purchaseCancelledError`
- SK2 / completion-block: `CustomerInfo` + cancelled + no error
- SK2 / `async`: `CustomerInfo` + cancelled (not error)

### After #1841:
- SK1 / completion-block:  no `CustomerInfo` + cancelled + `ErrorCode.purchaseCancelledError`
- SK1 / `async`: no `CustomerInfo` + ignored error -> crash! :rotating_light:
- SK2 / completion-block: `CustomerInfo` + cancelled + `ErrorCode.purchaseCancelledError`
- SK2 / `async`: `CustomerInfo` + cancelled (not error)

### Suggested change:
_To make SK1 and SK2 consistent and fix the crash on SK1_
_New behavior flagged with :warning:_

- SK1 / completion-block: `CustomerInfo` + cancelled + `ErrorCode.purchaseCancelledError`
- SK1 / `async`: `CustomerInfo` + cancelled (not error) :warning:
- SK2 / completion-block: `CustomerInfo` + cancelled + `ErrorCode.purchaseCancelledError` :warning:
- SK2 / `async`: `CustomerInfo` + cancelled (not error)

I think we should really make these consistent. My suggested result values I think are the ones that make the most sense:
- Completion block APIs: return a `CustomerInfo` and `cancelled`, but no error, since it's not really a "failed" condition
- `async` APIs: don't throw an error since it's not a "failed" condition

The downside is some users might be relying on a particular result right now, maybe assuming that if `CustomerInfo` is not `nil`  the purchase went through, or perhaps the same assumption if `error` is `nil`.
But considering the `PurchaseResultData` tuple and the completion-block values, I think this is what makes the most sense.

**Comment by NachoSoto** [comment]
I just submitted #1869. It doesn't fix the inconsistency but it fixes the crash for now.
If we agree on the new suggested behavior I can make another PR for that.

**Comment by NachoSoto** [comment]
@anivaros thanks for the report! The fix is now merged in `main` and should be released soon.
